src/charm.py

```diff
@@ -24,22 +28,80 @@ def __init__(self, *args: typing.Any) -> None:
             args: passthrough to CharmBase.
         """
         super().__init__(*args)
+        self._charm_state = CharmState(charm_config=self.config)
+        self._webserver = GunicornWebserver(charm_state=self._charm_state)
         self.framework.observe(self.on.config_changed, self.config_service)
 
+    def flask_container(self, require_connected: bool = True) -> Container:
+        """Get the flask application workload container controller.
+
+        Args:
+            require_connected: if set to ``True``, a runtime exception will be raised if the
+                pebble service inside the container is not ready.
+
+        Return:
+            The controller of the flask application workload container.
+
+        Raises:
+            RuntimeError: if the pebble service inside the container is not ready while the
+                ``require_connected`` is set to True.
+        """
+        container = self.unit.get_container("flask-app")
+        if require_connected and not container.can_connect():
+            raise RuntimeError("pebble inside flask-app container is not ready")
+        return container
+
     def config_service(self, event: ConfigChangedEvent) -> None:
         """Configure the flask pebble service layer.
 
         Args:
             event: the config-changed event that trigger this callback function.
         """
-        container = self.unit.get_container("flask-app")
+        container = self.flask_container(require_connected=False)
         if not container.can_connect():
             event.defer()
             return
+
+        service_name = "flask-app"
         container.add_layer("flask-app", self.flask_layer(), combine=True)
+        webserver_config_path = str(self._webserver.config_path)
+        current_webserver_config = self.pull_file(webserver_config_path)
```

Comment:
`pull_file` will return `None` if the file doesn't exist shouldn't we expect the file to be available at that time ? And start generating the file in the `install` event ?
This will avoid relying on the config_change event being triggered at startup

Reply:
If we generate the file during the install event then check the file in the config-changed event, that requires us to assume that install event happens before the config-changed event, which is exactly the kind of thing I would like to avoid.